hazelcast/include/hazelcast/client/connection/Connection.h

```diff
@@ -57,74 +63,107 @@ namespace hazelcast {
+                void onHeartbeatFailed();
 
-                std::auto_ptr<protocol::ClientMessage> readBlocking();
+                void onHeartbeatResumed();
 
-                bool isHeartBeating();
+                void onHeartbeatReceived();
 
-                void heartBeatingFailed();
+                void onHeartbeatRequested();
 
-                void heartBeatingSucceed();
+                bool isAuthenticatedAsOwner() const;
 
-                bool isOwnerConnection() const;
+                void setIsAuthenticatedAsOwner();
 
-                virtual void handleMessage(connection::Connection &connection, std::auto_ptr<protocol::ClientMessage> message);
+                virtual void handleClientMessage(const boost::shared_ptr<Connection> &connection,
+                                                 std::auto_ptr<protocol::ClientMessage> &message);
 
                 int getConnectionId() const;
 
-                void setConnectionId(int connectionId);
+                bool isAlive();
+
+                int64_t lastReadTimeMillis();
+
+                const std::string &getCloseReason() const;
+
+                void incrementPendingPacketCount();
+
+                void decrementPendingPacketCount();
+
+                int32_t getPendingPacketCount();
+
+                bool operator==(const Connection &rhs) const;
+
+                bool operator!=(const Connection &rhs) const;
+
+                const std::string &getConnectedServerVersionString() const;
+
+                void setConnectedServerVersion(const std::string &connectedServerVersionString);
+
+                std::auto_ptr<Address> getLocalSocketAddress() const;
+
+                int getConnectedServerVersion() const;
+
+                friend std::ostream &operator<<(std::ostream &os, const Connection &connection);
 
-                util::Atomic<time_t> lastRead;
-                util::AtomicBoolean live;
             private:
+                util::Atomic<int64_t> closedTimeMillis;
+                util::Atomic<int64_t> lastHeartbeatRequestedMillis;
+                // the time in millis the last heartbeat was received. 0 indicates that no heartbeat has ever been received.
+                util::Atomic<int64_t> lastHeartbeatReceivedMillis;
                 spi::ClientContext& clientContext;
-                spi::InvocationService& invocationService;
+                spi::ClientInvocationService &invocationService;
                 std::auto_ptr<Socket> socket;
                 ReadHandler readHandler;
                 WriteHandler writeHandler;
-                bool _isOwnerConnection;
+                bool authenticatedAsOwner;
                 util::AtomicBoolean heartBeating;
                 byte* receiveBuffer;
                 util::ByteBuffer receiveByteBuffer;
 
                 protocol::ClientMessageBuilder messageBuilder;
-                protocol::ClientMessage wrapperMessage;
                 std::auto_ptr<protocol::ClientMessage> responseMessage;
 
```

Comment:
After readBlocking is deleted most of the fields are seems not used. 
Followings can be deleted:
                 byte* receiveBuffer;	
                 util::ByteBuffer receiveByteBuffer;	
                 protocol::ClientMessageBuilder messageBuilder;	
                 std::auto_ptr<protocol::ClientMessage> responseMessage;

Reply:
Deleted as suggested.